internal/reconcilers/object_test.go

```diff
@@ -390,6 +390,17 @@ var _ = Describe("Basic propagation", func() {
 		Expect(objectInheritedFrom(ctx, "configmaps", barName, "foo-config")).Should(Equal(fooName))
 	})
 
+	It("should not be copied to descendents when source object is excluded", func() {
+		setParent(ctx, barName, fooName)
+		makeObject(ctx, "configmaps", fooName, "istio-ca-root-cert")
+		makeObject(ctx, "configmaps", fooName, "kube-root-ca.crt")
+		addToHNCConfig(ctx, "", "configmaps", api.Propagate)
+
+		// excluded ca cert configmaps should not be propagated from foo to bar.
+		Eventually(hasObject(ctx, "configmaps", barName, "istio-ca-root-cert")).Should(BeFalse())
+		Eventually(hasObject(ctx, "configmaps", barName, "kube-root-ca.crt")).Should(BeFalse())
```

Comment:
This could give a false positive even if the object _were_ propagated, because propagation takes time and this line could be executed before it completes.

A slightly better approach is to have an object that _is_ propagated, wait for that to be propagated _first_, and then use `Expect` instead of `Eventually` (since we're really not actually polling). You could also use `Consistently` which always polls _until_ the timeout completes, but TBH that's probably overkill here.

Whichever you pick, it's probably worth commenting too though since this is often not obvious to future devs!

Reply:
have added a propagated resource to this test and update these checks to `Expect` as suggested

Reply:
using `Expect` this test may periodically fail ref:  https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_hierarchical-namespaces/81/pull-hnc-test/1440412972798185472
so have switched back to `Eventually` but have added a configmap that is propagated to the test